### ISSUE TITLE
Remove Hive access entirely?

### ISSUE THREAD

**Issue by Altreus** [OpusVL/Flexibase#2]
It was suggested that we could follow a purer idea of dependency injection and remove access to the Hive from the Minds inside it.

Instead, each Mind will specify what services it needs and the Hive, on construction, will provide them.

Thus each Mind will never know about the Hive it is in, and simply gets given the instantiated services.

This works for services but leaves a question about the main feature of FB, which is to discover passive Hats providing extensions to arbitrary services.

Potentially each Mind can be given a Hat finding service but this seems hardly different from simply being given the Hive.

Potentially, the Hive could implement many services using Hats and as a result essentially have a Mind of its own.

**Comment by Nick-OpusVL** [comment]
> Potentially each Mind can be given a Hat finding service but this seems hardly different from simply being given the Hive.

Not quite.  We should look to Django's signals framework I think, as an example of how you can have a signal that is "global"[1], can be passed to its 'emitter', subscribed to by its subscribers (by means of side effects in constructors), and *the emitter code can gather the return values from the subscribers' event handlers*.  So for example an emitter could be dbicdh::manager and it would gather all the schemas that its signal's subscribers returned and then work on them all when they've all returned.

[1] by "global" I mean can be global - it shouldn't be hard-wired to be a singleton if we can possibly avoid it

**Comment by Altreus** [comment]
Seems like the fundamentals of Flexibase have become: it turns out you can use an event system as a discovery mechanism.

In any event system, if you create "Can anyone *verb*?" as an event, then you can collect respondents and call something on all of them. If you construct it properly, you can arrange it so that for some values of *verb* you only have one respondent, and for others you have many: the difference between a service and passive hats.

What nuances have I glossed over here?
